code/__HELPERS/path.dm

```diff
@@ -0,0 +1,346 @@
+	/// Do we only worry about turfs with simulated atmos, most notably things that aren't space?
+	var/simulated_only
+	/// Did we succeed?
+	var/success = FALSE
+
+/datum/pathfind/New(atom/movable/caller, atom/goal, id, maxnodes, maxnodedepth, mintargetdist, simulated_only)
+	src.caller = caller
+	end = get_turf(goal)
+	open = new /datum/heap(/proc/HeapPathWeightCompare)
+	openc = new() //open list for node check
+	sources = new() //open list for node check
+	src.id = id
+	src.maxnodes = maxnodes
+	src.maxnodedepth = maxnodedepth
+	src.mintargetdist = mintargetdist
+	src.simulated_only = simulated_only
+
+//datum/pathfind/proc/generate_node(turf/the_tile, )
+/datum/pathfind/proc/start_search()
+	caller.calculating_path = TRUE
+	start = get_turf(caller)
+	if(!start || !end)
+		stack_trace("Invalid A* start or destination")
+		return FALSE
+	if(start.z != end.z || start == end ) //no pathfinding between z levels
+		return FALSE
+	if(maxnodes) //if start turf is farther than maxnodes from end turf, no need to do anything
+		maxnodedepth = maxnodes
+
+	//initialization
+	var/datum/jpsnode/cur = new (start,null,0,end)//current processed turf
+	open.Insert(cur)
+	openc[start] = cur
+	sources[start] = PATH_START
+	//then run the main loop
+	var/total_tiles
+
+	while(!open.IsEmpty() && !path)
+		if(!caller)
+			return
+		cur = open.Pop() //get the lower f turf in the open list
+
+		total_tiles++
+		if((maxnodedepth)&&(cur.nt > maxnodedepth))//if too many steps, don't process that path
+			continue
+
+		var/turf/current_turf = cur.tile
+
+		for(var/scan_direction in list(NORTH, EAST, SOUTH, WEST))
+			lateral_scan_spec(current_turf, scan_direction) // this is a turf not a node, fix
+			if(path)
+				break
+
+		for(var/scan_direction in list(NORTHEAST, SOUTHEAST, SOUTHWEST, NORTHWEST))
+			diag_scan_spec(current_turf, scan_direction)
+			if(path)
+				break
+
+		CHECK_TICK
+	//reverse the path to get it from start to finish
+	if(path)
+		for(var/i = 1 to round(0.5*path.len))
+			path.Swap(i,path.len-i+1)
+	//testing("*********new path done with [total_tiles] tiles popped, [openc.len] nodes added to heap")
+	openc = null
+	//cleaning after us
+	caller.calculating_path = FALSE
+	//return path
+	return total_tiles
+
+///
+/datum/pathfind/proc/unwind_path(datum/jpsnode/unwind_node)
+	success = unwind_node.tile == end
+	path = new()
+	var/turf/iter_turf = unwind_node.tile
+	var/turf/iter_turf2 = get_turf(unwind_node.tile)
+	path.Add(iter_turf)
+
+	var/list/b = list(iter_turf2)
+	var/turf/goal_turf2 = sources[iter_turf2]
+	while(goal_turf2 && goal_turf2 != PATH_START)
+		b.Add(goal_turf2)
+		while(iter_turf2 != goal_turf2)
+			var/turf/next_turf = get_step_towards(iter_turf2, goal_turf2)
+			iter_turf2 = next_turf
+			path.Add(iter_turf2)
+		goal_turf2 = sources[iter_turf2]
+
+	//testing("LENGTHS OF 2 LISTS| A: [a.len] B: [b.len] (final goalturf2 = [goal_turf2])")
+
+/datum/pathfind/proc/can_step(turf/a, turf/next) // prolly not optimal
```

Comment:
at least use define here, this is most common procall here, shouldn't be wrapped like this.

Reply:
+1, this should be a macro

Reply:
Any reason to use `call()`? It carries its own overhead.

Reply:
In theory this function can be variable but we only use 2? variants for this so it might as well be a switch.